Modified roundup_power2_divisions to specify the number of divisions for each power of two interval

c-odrin opened the pull request:
Summary:
Improved roundup_power2_divisions knob so it allows better control of rouding in the PyTorch CUDA Caching Allocator.

This new version allows setting the number of divisions per power of two interval starting from 1MB and ending at 64GB and above. An example use case is when rouding is desirable for small allocations but there are also very large allocations which are persistent, thus would not benefit from rounding and take up extra space.

Test Plan: Tested locally

Differential Revision: D40103909

zdevito reviewed (commented):
The API looks good, thanks for making the configuration string short and understandable.  I have a couple inline comments. Before we can merge this we will also need a couple tests that this actually rounds as expected, see `tests/test_cuda.py` for other allocated related tests where the config string is set and we look to see if the allocations made are rounded as expected.

c-odrin commented:
Thank you for the comments @zdevito. I have now added tests in `tests/test_cuda.py` and changed the implementation to not allow incorrectly formatted config strings. The other two inline comments were also addressed.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -423,128 +437,267 @@ class CachingAllocatorConfig {
+    const char* env,
+    std::vector<std::string>& config) {
+  std::string buf;
+  for (size_t i = 0; i < strlen(env); i++) {
```
nit: `strlen` is O(n) so should be lifted out of the loop

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -423,128 +437,267 @@ class CachingAllocatorConfig {
+void CachingAllocatorConfig::lexArgs(
+    const char* env,
+    std::vector<std::string>& config) {
+  std::string buf;
```
`std::vector<char>` to avoid all the string allocation and contacting.

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -423,128 +437,267 @@ class CachingAllocatorConfig {
+      }
+      config.emplace_back(std::string(1, env[i]));
+    } else if (
+        isalpha(env[i]) || isdigit(env[i]) || env[i] == '>' || env[i] == '_' ||
```
To make this easier for new options int he future, and since the tokens have to be checked for validity later anyway:
```
else if (env[i] != ' ') {
  buf.emplace_back(env[i])
}
```

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -423,128 +437,267 @@ class CachingAllocatorConfig {
+
+  for (size_t i = 0; i < config.size(); i++) {
+    if (config[i].compare("max_split_size_mb") == 0) {
+      parseMaxSplitSize(config, i);
```
passing `i` by mutable reference here is surprising. Maybe either make it a pointer to it is clear it is being mutated, or return the new position.

zdevito reviewed (commented):
This looks good. I just have a couple minor things in the parsing. Thanks!

c-odrin commented:
Thank you for the comments! I've addressed them in the newest version

c-odrin committed:
Modified roundup_power2_divisions to specify the number of divisions for each power of two interval (#87290)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/87290

Improved roundup_power2_divisions knob so it allows better control of rouding in the PyTorch CUDA Caching Allocator.

This new version allows setting the number of divisions per power of two interval starting from 1MB and ending at 64GB and above. An example use case is when rouding is desirable for small allocations but there are also very large allocations which are persistent, thus would not benefit from rounding and take up extra space.

Example setting this knob:
To set 1 division for all allocations under 256MB, 2 division for allocations between 256MB and 512MB, 4 divisions for allocations between 512MB and 1GB and 8 divisions for any larger allocations, set the knob value to:

```
roundup_power2_divisions:[256:1,512:2,1024:4,>:8].
```

Test Plan:
Added testcase to caffe2/test/test_cuda.py (run as part of GitHub tests)

Tested parser locally using the following sample program: P545400980

Reviewed By: jaewonlee-fb

Differential Revision: D40103909

fbshipit-source-id: 2a4b69cbd2cad62f185ff5cb5edc9bf95f075983

zdevito reviewed (approved):
Changes look good! Thanks
